### ISSUE TITLE
pageserver: simplified fsync logic for tenant directories

### ISSUE THREAD

**Issue by jcsp** [neondatabase/neon#6041]
General issue: the data in a tenant directory is predominantly a disposable cache.  However, there are a couple of places we really care about atomic creation of files:
- `timelines/`, where after a restart we expect to be able to e.g. create timeline dirs without each dir creation having to optionally create the parent timelines/.
- Tenant configuration files, where the existence of the configuration file is what prompts the pageserver to re-attach on restart, so we should not respond to attach HTTP requests until we have ensured durability of this file and its ancestors.

Currently we use `crashsafe_overwrite` for configs and `crashsafe::create_dir` for the timelines/ dir.  This is doing more system calls than are really required.  The order should be:
- Create all directories from root to timelines/
- Do a crashsafe_overwrite of the tenant conf (this includes fsync'ing the tenant dir, therefore implicitly ensures durability of the timelines/ dir we created).
- fsync the parent of the tenant dir to ensure its durability.

It is not necessary to atomically create the tenants dir and the config file.  If a tenant dir is created with no config file.

Then, on restart, we are guaranteed that:
- If an attach operation responded to the caller, then the tenant conf will be present.
- If the tenant conf is present, then the timelines dir is also present (we created it before we created the tenant conf file).

**Comment by knizhnik** [comment]
I first time heard about this idea:  fully ephemeral pageservers.
Were there any discussion on it?

IMHO it is not realistic. Certainly removing fsync overhead can improve PS performance. But retrieving layers from S3 requires too much time and so cause unacceptable slowdown of client applications. If we do not have sync, then we can not guarantee consistent state of PS. It means that each abnormal PS termination should cause restart PS with empty storage: all projects have to be downloaded from S3.

**Comment by jcsp** [comment]
>I first time heard about this idea: fully ephemeral pageservers.

Yes, the storage team has discussed this in detail.  It doesn't mean that we would drop all layers on restart, it just means that the pageserver's on-disk state wouldn't be used as a source of truth for what should be attached on restart.

> It means that each abnormal PS termination should cause restart PS with empty storage: all projects have to be downloaded from S3.

No, this isn't the case.  A fsync-less pageserver can recover after restart by dropping any incomplete layer files -- we intended to add checksums for layers as a general correctness check, and those would enable the pageserver to distinguish a valid or invalid layer file.

**Comment by knizhnik** [comment]
Hmmm... reading all layers on restart for calculation checksums seems to be also not so good idea...

**Comment by jcsp** [comment]
> Hmmm... reading all layers on restart for calculation checksums seems to be also not so good idea...

We'd do it lazily, not in one shot at start.  In any case, we're well off topic for this issue: the scope of this issue is just the way we fsync tenant directories on create/attach/restart, not the management of layer files.

**Comment by knizhnik** [comment]
Ok, sorry that I once again start the discussion in the unrelated ticket.
But as far as I didn't see any other RFC or ticket where it was discussed - I did it where I first heard about this proposal.

My last comment:
- Fsyncing large files (and 128Mb  layer file is definitely such one) is not is expensive. I mean that if we in any case what to persist this data, waiting completion of this operation should not cause significant degrade of performance. And may have even positive effect, repenting so called "write storm", when OS has to flush huge number of dirty buffers.
- Certainly I understand that we will not calculate CRC for all layers - just for one which is retrieved by clients. Actually logical size calculation performed at startup may trigger loading large number of layers.  Yes, it is now delayed. But has to be done some times in any case. But even if we take in account only layers directly requested by client's `get_page` requests, then calculating CRC for 128Mb image may be quite expensive. At least I suggest to perform some experiments beforetrying to implement this ephemeral page server model.
